pkg/option/option.go

```diff
@@ -0,0 +1,127 @@
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+package o
+
+import "fmt"
+
+// Option represents an optional value. The [Some] variant contains a value and
+// the [None] variant represents the absence of a value.
+type Option[T any] struct {
+	value   T
+	present bool
+}
+
+// Some instantiates an [Option] with a value.
+func Some[T any](value T) Option[T] {
+	return Option[T]{value, true}
+}
+
+// None instantiates an [Option] with no value.
+func None[T any]() Option[T] {
+	return Option[T]{}
+}
+
+// String implements the [fmt.Stringer] interface.
+func (o Option[T]) String() string {
+	if o.present {
+		return fmt.Sprintf("Some(%v)", o.value)
+	}
+
+	return "None"
+}
+
+var _ fmt.Stringer = Option[struct{}]{}
+
+// Unwrap returns the underlying value of a [Some] variant, or panics if called
+// on a [None] variant.
+func (o Option[T]) Unwrap() T {
+	if o.present {
+		return o.value
+	}
+
+	panic("called `Option.Unwrap()` on a `None` value")
+}
+
+// UnwrapOr returns the underlying value of a [Some] variant, or the provided
+// value on a [None] variant.
+func (o Option[T]) UnwrapOr(value T) T {
+	if o.present {
+		return o.value
+	}
+
+	return value
+}
+
+// UnwrapOrElse returns the underlying value of a [Some] variant, or the result
+// of calling the provided function on a [None] variant.
+func (o Option[T]) UnwrapOrElse(f func() T) T {
+	if o.present {
+		return o.value
+	}
+
+	return f()
+}
+
+// UnwrapOrZero returns the underlying value of a [Some] variant, or the zero
+// value on a [None] variant.
+func (o Option[T]) UnwrapOrZero() T {
+	if o.present {
+		return o.value
+	}
+
+	var value T
+	return value
+}
+
+// IsSome returns true if the [Option] is a [Some] variant.
+func (o Option[T]) IsSome() bool {
+	return o.present
+}
+
+// IsNone returns true if the [Option] is a [None] variant.
+func (o Option[T]) IsNone() bool {
+	return !o.present
+}
+
+// Value returns the underlying value and true for a [Some] variant, or the
+// zero value and false for a [None] variant.
+func (o Option[T]) Value() (T, bool) {
+	return o.value, o.present
+}
+
+// Expect returns the underlying value for a [Some] variant, or panics with the
+// provided message for a [None] variant.
+func (o Option[T]) Expect(message string) T {
+	if o.present {
+		return o.value
+	}
+
+	panic(message)
+}
+
+func Map[T, U any](o Option[T], f func(T) U) Option[U] {
```

Comment:
What would a comment describing `Map` say?

```suggestion
// Map returns the transformed result of the underlying value of a [Some] variant through the provided function, or a [None] variant.
func Map[T, U any](o Option[T], f func(T) U) Option[U] {
```

Reply:
I think what you've written is fine. Rust docs write:

> Maps an Option<T> to Option<U> by applying a function to a contained value (if Some) or returns None (if None).

Reply:
That works, too.  Both read a bit like stereo instructions, requiring extra effort to look at how its used. 😅

Reply:
Yeh I hear you, it's kinda unintelligible. It's probably easier to look at the [doc examples or tests.](https://github.com/cli/cli/blob/7eddd537ba74e90ea141fe2f4daadbf6b9d01273/pkg/option/option_test.go#L102-L109). All that said, I don't think it's that much less clear than `Array.map`; what does mapping even mean? When you know you just kinda know.

Count yourself lucky we're not dealing in Haskell here otherwise you get magic like:

> A type f is a Functor if it provides a function fmap which, given any types a and b, lets you apply any function of type (a -> b) to turn an f a into an f b, preserving the structure of f. 🙅